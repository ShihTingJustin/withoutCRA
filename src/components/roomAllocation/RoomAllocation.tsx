import React, { useEffect, useState, useRef } from 'react';

import CustomInputNumber from '../customInputNumber/CustomInputNumber';

import './roomAllocation.scss';

type TotalGuest = { adult: number; child: number };

const adultMinimum = 1;

const Room = ({
  guestLimit,
  yetDistributedCount,
  disabled,
  onChange
}: {
  guestLimit: number;
  yetDistributedCount: number;
  disabled?: boolean;
  minusDisabled?: boolean;
  plusDisabled?: boolean;
  onChange: (value: TotalGuest) => void;
}) => {
  const [guestCount, setGuestCount] = useState({ adult: adultMinimum, child: 0 });

  const guestCountRef = useRef(0);

  const roomDisabled = yetDistributedCount === 0 || guestCountRef.current === guestLimit;

  const handleChange = (guestType: string, value: number) => {
    setGuestCount((prev) => ({ ...prev, [guestType]: value }));
  };

  useEffect(() => {
    onChange(guestCount);
    guestCountRef.current = Object.values(guestCount).reduce((acc, curr) => acc + curr);
  }, [guestCount]);

  return (
    <div className="roomAllocation-root__guest-input-wrap__room">
      <div className="counter">房間 : {guestCount.adult + guestCount.child} 人</div>
      <div className="row">
        <div className="guest-type">
          <p>大人</p>
          <span>年齡 20+</span>
        </div>
        <CustomInputNumber
          name="CustomInputNumber"
          max={guestLimit - guestCount.child}
          min={adultMinimum}
          value={guestCount.adult}
          disabled={disabled}
          plusDisabled={roomDisabled}
          yetDistributedCount={yetDistributedCount}
          onBlur={(e) => console.log(e)}
          onChange={(e) => {
            handleChange('adult', Number(e.target.value));
          }}
        />
      </div>
      <div className="row">
        <div className="guest-type">
          <p>小孩</p>
        </div>
        <CustomInputNumber
          name="CustomInputNumber"
          max={guestLimit - guestCount.adult}
          min={0}
          value={guestCount.child}
          disabled={disabled}
          plusDisabled={roomDisabled}
          yetDistributedCount={yetDistributedCount}
          onBlur={(e) => console.log(e)}
          onChange={(e) => {
            handleChange('child', Number(e.target.value));
          }}
        />
      </div>
    </div>
  );
};

const RoomAllocation = ({
  guest,
  room,
  guestLimit,
  onChange
}: {
  guest: number;
  room: number;
  guestLimit: number;
  onChange: (result: TotalGuest[]) => void;
}) => {
  const result = useRef<TotalGuest[]>([]);

  const [yetDistributedCount, setYetDistributedCount] = useState<number>(guest);

  const handleChange = (index: number, value: TotalGuest) => {
    // get data for parent onChange callback
    if (!result.current[index]) {
      result.current[index] = { adult: 0, child: 0 };
    }
    result.current[index] = value;
    onChange(result.current);

    // calculate totalGuestCount then update yetDistributedCount
    const totalGuestCount = result.current.reduce((total, singleRoom) => {
      const roomCount = Object.values(singleRoom).reduce((acc, curr) => acc + curr);
      return (total += roomCount);
    }, 0);
    setYetDistributedCount(guest - totalGuestCount);
  };

  return (
    <div className="roomAllocation-root">
      <div className="roomAllocation-root__room-info-wrap">
        <div className="total-guest-room">
          住客人數 : {guest} 人 / {room} 房
        </div>
        <div className="yet-distribute-counter">尚未分配人數 : {yetDistributedCount} 人</div>
      </div>
      <div className="roomAllocation-root__guest-input-wrap">
        {Array.from({ length: room }, (_, index) => (
          <Room
            key={index}
            guestLimit={guestLimit}
            yetDistributedCount={yetDistributedCount}
            disabled={guest === room}
            onChange={(value) => handleChange(index, value)}
          />
        ))}
      </div>
    </div>
  );
};

export default RoomAllocation;
