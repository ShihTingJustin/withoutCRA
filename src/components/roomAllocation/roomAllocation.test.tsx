import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';

import RoomAllocation from './RoomAllocation';

describe('Test Button component', () => {
  const component = <RoomAllocation guest={10} room={3} guestLimit={4} />;

  it('[Render] match snapshot', () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('[Render] toBeInTheDocument', () => {
    render(component);
    const element = screen.getByTestId('roomAllocation');
    expect(element).toBeInTheDocument();
  });
});
