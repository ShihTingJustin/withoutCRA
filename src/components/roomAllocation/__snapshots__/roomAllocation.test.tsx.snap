// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Button component [Render] match snapshot 1`] = `
<div
  className="roomAllocation-root"
  data-testid="roomAllocation"
>
  <div
    className="roomAllocation-root__room-info-wrap"
  >
    <div
      className="total-guest-room"
    >
      住客人數 : 
      10
       人 / 
      3
       房
    </div>
    <div
      className="yet-distribute-counter"
    >
      尚未分配人數 : 
      10
       人
    </div>
  </div>
  <div
    className="roomAllocation-root__guest-input-wrap"
  >
    <div
      className="roomAllocation-root__guest-input-wrap__room"
    >
      <div
        className="counter"
      >
        房間 : 
        1
         人
      </div>
      <div
        className="row"
      >
        <div
          className="guest-type"
        >
          <p>
            大人
          </p>
          <span>
            年齡 20+
          </span>
        </div>
        <div
          className="input-number-root"
          data-testid="customInputNumber"
        >
          <div
            className="button-root"
            data-testid="button"
          >
            <button
              data-disabled={false}
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
            >
              <span>
                <svg
                  color="#1e9fd2"
                  fill="currentColor"
                  height={30}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    {
                      "color": "#1e9fd2",
                    }
                  }
                  viewBox="0 0 1024 1024"
                  width={30}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M872 474H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h720c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <input
            data-disabled={false}
            data-testid="customInputNumber-input"
            disabled={false}
            name="CustomInputNumber"
            onBlur={[Function]}
            onChange={[Function]}
            value={1}
          />
          <div
            className="button-root"
            data-testid="button"
          >
            <button
              data-disabled={false}
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
            >
              <span>
                <svg
                  color="#1e9fd2"
                  fill="currentColor"
                  height={30}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    {
                      "color": "#1e9fd2",
                    }
                  }
                  t="1551322312294"
                  version="1.1"
                  viewBox="0 0 1024 1024"
                  width={30}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <defs />
                  <path
                    d="M474 152m8 0l60 0q8 0 8 8l0 704q0 8-8 8l-60 0q-8 0-8-8l0-704q0-8 8-8Z"
                  />
                  <path
                    d="M168 474m8 0l672 0q8 0 8 8l0 60q0 8-8 8l-672 0q-8 0-8-8l0-60q0-8 8-8Z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="guest-type"
        >
          <p>
            小孩
          </p>
        </div>
        <div
          className="input-number-root"
          data-testid="customInputNumber"
        >
          <div
            className="button-root"
            data-testid="button"
          >
            <button
              data-disabled={false}
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
            >
              <span>
                <svg
                  color="#1e9fd2"
                  fill="currentColor"
                  height={30}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    {
                      "color": "#1e9fd2",
                    }
                  }
                  viewBox="0 0 1024 1024"
                  width={30}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M872 474H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h720c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <input
            data-disabled={false}
            data-testid="customInputNumber-input"
            disabled={false}
            name="CustomInputNumber"
            onBlur={[Function]}
            onChange={[Function]}
            value={0}
          />
          <div
            className="button-root"
            data-testid="button"
          >
            <button
              data-disabled={false}
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
            >
              <span>
                <svg
                  color="#1e9fd2"
                  fill="currentColor"
                  height={30}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    {
                      "color": "#1e9fd2",
                    }
                  }
                  t="1551322312294"
                  version="1.1"
                  viewBox="0 0 1024 1024"
                  width={30}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <defs />
                  <path
                    d="M474 152m8 0l60 0q8 0 8 8l0 704q0 8-8 8l-60 0q-8 0-8-8l0-704q0-8 8-8Z"
                  />
                  <path
                    d="M168 474m8 0l672 0q8 0 8 8l0 60q0 8-8 8l-672 0q-8 0-8-8l0-60q0-8 8-8Z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="roomAllocation-root__guest-input-wrap__room"
    >
      <div
        className="counter"
      >
        房間 : 
        1
         人
      </div>
      <div
        className="row"
      >
        <div
          className="guest-type"
        >
          <p>
            大人
          </p>
          <span>
            年齡 20+
          </span>
        </div>
        <div
          className="input-number-root"
          data-testid="customInputNumber"
        >
          <div
            className="button-root"
            data-testid="button"
          >
            <button
              data-disabled={false}
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
            >
              <span>
                <svg
                  color="#1e9fd2"
                  fill="currentColor"
                  height={30}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    {
                      "color": "#1e9fd2",
                    }
                  }
                  viewBox="0 0 1024 1024"
                  width={30}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M872 474H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h720c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <input
            data-disabled={false}
            data-testid="customInputNumber-input"
            disabled={false}
            name="CustomInputNumber"
            onBlur={[Function]}
            onChange={[Function]}
            value={1}
          />
          <div
            className="button-root"
            data-testid="button"
          >
            <button
              data-disabled={false}
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
            >
              <span>
                <svg
                  color="#1e9fd2"
                  fill="currentColor"
                  height={30}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    {
                      "color": "#1e9fd2",
                    }
                  }
                  t="1551322312294"
                  version="1.1"
                  viewBox="0 0 1024 1024"
                  width={30}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <defs />
                  <path
                    d="M474 152m8 0l60 0q8 0 8 8l0 704q0 8-8 8l-60 0q-8 0-8-8l0-704q0-8 8-8Z"
                  />
                  <path
                    d="M168 474m8 0l672 0q8 0 8 8l0 60q0 8-8 8l-672 0q-8 0-8-8l0-60q0-8 8-8Z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="guest-type"
        >
          <p>
            小孩
          </p>
        </div>
        <div
          className="input-number-root"
          data-testid="customInputNumber"
        >
          <div
            className="button-root"
            data-testid="button"
          >
            <button
              data-disabled={false}
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
            >
              <span>
                <svg
                  color="#1e9fd2"
                  fill="currentColor"
                  height={30}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    {
                      "color": "#1e9fd2",
                    }
                  }
                  viewBox="0 0 1024 1024"
                  width={30}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M872 474H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h720c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <input
            data-disabled={false}
            data-testid="customInputNumber-input"
            disabled={false}
            name="CustomInputNumber"
            onBlur={[Function]}
            onChange={[Function]}
            value={0}
          />
          <div
            className="button-root"
            data-testid="button"
          >
            <button
              data-disabled={false}
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
            >
              <span>
                <svg
                  color="#1e9fd2"
                  fill="currentColor"
                  height={30}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    {
                      "color": "#1e9fd2",
                    }
                  }
                  t="1551322312294"
                  version="1.1"
                  viewBox="0 0 1024 1024"
                  width={30}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <defs />
                  <path
                    d="M474 152m8 0l60 0q8 0 8 8l0 704q0 8-8 8l-60 0q-8 0-8-8l0-704q0-8 8-8Z"
                  />
                  <path
                    d="M168 474m8 0l672 0q8 0 8 8l0 60q0 8-8 8l-672 0q-8 0-8-8l0-60q0-8 8-8Z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="roomAllocation-root__guest-input-wrap__room"
    >
      <div
        className="counter"
      >
        房間 : 
        1
         人
      </div>
      <div
        className="row"
      >
        <div
          className="guest-type"
        >
          <p>
            大人
          </p>
          <span>
            年齡 20+
          </span>
        </div>
        <div
          className="input-number-root"
          data-testid="customInputNumber"
        >
          <div
            className="button-root"
            data-testid="button"
          >
            <button
              data-disabled={false}
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
            >
              <span>
                <svg
                  color="#1e9fd2"
                  fill="currentColor"
                  height={30}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    {
                      "color": "#1e9fd2",
                    }
                  }
                  viewBox="0 0 1024 1024"
                  width={30}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M872 474H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h720c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <input
            data-disabled={false}
            data-testid="customInputNumber-input"
            disabled={false}
            name="CustomInputNumber"
            onBlur={[Function]}
            onChange={[Function]}
            value={1}
          />
          <div
            className="button-root"
            data-testid="button"
          >
            <button
              data-disabled={false}
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
            >
              <span>
                <svg
                  color="#1e9fd2"
                  fill="currentColor"
                  height={30}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    {
                      "color": "#1e9fd2",
                    }
                  }
                  t="1551322312294"
                  version="1.1"
                  viewBox="0 0 1024 1024"
                  width={30}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <defs />
                  <path
                    d="M474 152m8 0l60 0q8 0 8 8l0 704q0 8-8 8l-60 0q-8 0-8-8l0-704q0-8 8-8Z"
                  />
                  <path
                    d="M168 474m8 0l672 0q8 0 8 8l0 60q0 8-8 8l-672 0q-8 0-8-8l0-60q0-8 8-8Z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="guest-type"
        >
          <p>
            小孩
          </p>
        </div>
        <div
          className="input-number-root"
          data-testid="customInputNumber"
        >
          <div
            className="button-root"
            data-testid="button"
          >
            <button
              data-disabled={false}
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
            >
              <span>
                <svg
                  color="#1e9fd2"
                  fill="currentColor"
                  height={30}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    {
                      "color": "#1e9fd2",
                    }
                  }
                  viewBox="0 0 1024 1024"
                  width={30}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M872 474H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h720c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <input
            data-disabled={false}
            data-testid="customInputNumber-input"
            disabled={false}
            name="CustomInputNumber"
            onBlur={[Function]}
            onChange={[Function]}
            value={0}
          />
          <div
            className="button-root"
            data-testid="button"
          >
            <button
              data-disabled={false}
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
            >
              <span>
                <svg
                  color="#1e9fd2"
                  fill="currentColor"
                  height={30}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    {
                      "color": "#1e9fd2",
                    }
                  }
                  t="1551322312294"
                  version="1.1"
                  viewBox="0 0 1024 1024"
                  width={30}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <defs />
                  <path
                    d="M474 152m8 0l60 0q8 0 8 8l0 704q0 8-8 8l-60 0q-8 0-8-8l0-704q0-8 8-8Z"
                  />
                  <path
                    d="M168 474m8 0l672 0q8 0 8 8l0 60q0 8-8 8l-672 0q-8 0-8-8l0-60q0-8 8-8Z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
